daemon/cluster/controllers/plugin/controller.go

```diff
@@ -1,79 +1,264 @@
 package plugin
 
 import (
+	"encoding/base64"
+	"encoding/json"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"strings"
+
 	"github.com/Sirupsen/logrus"
+	"github.com/docker/distribution/reference"
+	enginetypes "github.com/docker/docker/api/types"
+	"github.com/docker/docker/api/types/swarm/runtime"
+	"github.com/docker/docker/plugin"
+	"github.com/docker/docker/plugin/v2"
 	"github.com/docker/swarmkit/api"
+	"github.com/gogo/protobuf/proto"
+	"github.com/pkg/errors"
 	"golang.org/x/net/context"
 )
 
-// Controller is the controller for the plugin backend
-type Controller struct{}
+// Controller is the controller for the plugin backend.
+// Plugins are managed as a singleton object with a desired state (different from containers).
+// With the the plugin controller instead of having a strict create->start->stop->remove
+// task lifecycle like containers, we manage the desired state of the plugin and let
+// the plugin manager do what it already does and monitor the plugin.
+// We'll also end up with many tasks all pointing to the same plugin ID.
```

Comment:
I didn't fully understand this. Does it mean that instead of one task per plugin per node, every action of installing a plugin, enabling a plugin, removing a plugin, etc., gets its own task?

Does this require any orchestration changes in swarmkit, or is it accomplished by creating a new global service for each operation?

What's the strategy for garbage collecting old tasks and/or services that are no longer relevant?